test/e2e/framework/framework.go

```diff
@@ -0,0 +1,295 @@
+    namespacesToDelete       []*v1.Namespace // Some tests have more than one.
+    NamespaceDeletionTimeout time.Duration
+
+    // To make sure that this framework cleans up after itself, no matter what,
+    // we install a Cleanup action before each test and clear it after.  If we
+    // should abort, the AfterSuite hook should run all Cleanup actions.
+    cleanupHandle            CleanupActionHandle
+
+    // configuration for framework's client
+    Options Options
+
+}
+
+// Options is a struct for managing test framework options.
+type Options struct {
+    ClientQPS    float32
+    ClientBurst  int
+    GroupVersion *schema.GroupVersion
+}
+
+// NewDefaultFramework makes a new framework and sets up a BeforeEach/AfterEach for
+// you (you can write additional before/after each functions).
+func NewDefaultFramework(baseName string) *Framework {
+    options := Options{
+        ClientQPS:   20,
+        ClientBurst: 50,
+    }
+    return NewFramework(baseName, options)
+}
+
+// NewFramework creates a test framework.
+func NewFramework(baseName string, options Options) *Framework {
+    f := &Framework{
+        BaseName:                 baseName,
+        Options:                  options,
+    }
+
+    ginkgo.BeforeEach(f.BeforeEach)
+    ginkgo.AfterEach(f.AfterEach)
+
+    return f
+}
+
+
+func (f *Framework) BeforeEach() {
+    // workaround for a bug in ginkgo.
+    // https://github.com/onsi/ginkgo/issues/222
+    f.cleanupHandle = AddCleanupAction(f.AfterEach)
+
+    ginkgo.By("Creating kubernetes clients")
+    for _, context := range TestContext.KubeContexts {
+        client := f.createKubernetesClient(context)
+        f.ClusterClients = append(f.ClusterClients, client)
+    }
+
+    if !f.SkipNamespaceCreation {
+        ginkgo.By(fmt.Sprintf("Building namespace api objects, basename %s", f.BaseName))
+
+        namespaceLabels := map[string]string {
+            "e2e-framework": f.BaseName,
+        }
+
+        for idx, clientSet := range f.ClusterClients {
+            switch idx {
+            case ClusterA: // On the first cluster we let k8s generate a name for the namespace
+                namespace := generateNamespace(clientSet, f.BaseName, namespaceLabels)
+                f.Namespace = namespace.GetName()
+                f.UniqueName = namespace.GetName()
+            default: // On the other clusters we use the same name to make tracing easier
+                f.CreateNamespace(clientSet, f.Namespace, namespaceLabels)
+            }
+        }
+    } else {
+        f.UniqueName = string(uuid.NewUUID())
+    }
+
+}
+
+func (f *Framework) createKubernetesClient(context string) *kubeclientset.Clientset {
+
+    restConfig, _, err := loadConfig(TestContext.KubeConfig, context)
+    if err != nil {
+        Errorf("Unable to load kubeconfig file %s for context %s, this is a non-recoverable error",
+            TestContext.KubeConfig, context)
+        Errorf("loadConfig err: %s", err.Error())
+        os.Exit(1)
```

Comment:
I think we should be consistent and fail the test via assertion as we do elsewhere instead of via Exit, ie

    Expect(err).NotTo(HaveOccurred(), "Unable to load kubeconfig file %s for context %s", 
        TestContext.KubeConfig, context)

Reply:
In this case I believe we should just exit testing and stop everything. Otherwise we will get "100" tests failing, where, the issue is not the specific case being tested to be failed, but miss configuration.

So in this case, we can tell who executed the test: hey, your config file is wrong, or missing, please fix it, and I will run the tests.

At this point, nothing has been ran in kubernetes, so there's nothing to cleanup.

Reply:
ok good point - seems reasonable.